### ISSUE TITLE
Feedforward with rounded activation values and Backprop with unrounded activation values

### ISSUE THREAD

**Issue by nhannguyen95** [keras-team/keras#9396]
Hi, the activation values of hidden layers in my Neural Net are in [0, 1]. 

At hidden layer i: let's say its activation values are x, and round(x) is the rounded value of x (only 0, 1)
+ I want to assign round(x) to this hidden layer, and feedforward it to the later layers.
+ However, when doing Backprop, the Neural Net will use x (instead of round(x)).

The question is how can I implement this in Keras?
Thank you in advance.

**Comment by kaichi040696** [comment]
Same issue as mine. Can anyone answer this question?

**Comment by nhannguyen95** [comment]
Hi @kaichi040696 , I still don't know how to implement this in Keras. However, I managed to code this in Tensorflow, hope you find it helpful.

The trick here is: if you want to calculate y = f(x) in feedforward, and compute gradient with g(x) in backprop, at node y use:

`y = g(x) + tf.stop_gradient(f(x) - g(x))`

which means:
- At feedforward, y = g + f - g = f.
- At backprop, tf.stop_gradient operator stops the chain rule at (f(x) - g(x)). Now dy/dx = dg/dx.

An example in practice: for example at the tensor you want to perform the rounded/unrounded calculation:

`
current_unrounded_node = tf.layers.dense(previous_node, ...)
current_rounded_node = tf.round(unrounded_node)

next_unrounded_node = fully_connected(current_unrounded_node)
next_rounded_node = fully_connected(current_rounded_node)
next_node = next_unrounded_node 
                   + tf.stop_gradient(next_rounded_node - next_unrounded_node)

# continue the computation with next_node
`

Basically `tf.layers.dense` and `fully_connected` do the same thing (perform a fully connected layer). But since we need `next_unrounded_node` and `next_rounded_node` are computed from a same set of weight and bias, we should have a customized version of it to control the weights. For example:

`
def fully_connected(inputs, out_dim, scope_name, activation='relu'):
    with tf.variable_scope(scope_name, reuse=tf.AUTO_REUSE) as scope:
        in_dim = inputs.shape[-1]
        w = tf.get_variable('weights', [in_dim, out_dim],
                            initializer=tf.truncated_normal_initializer())
        b = tf.get_variable('bias', [out_dim],
                            initializer=tf.constant_initializer(0.0))
        out = tf.matmul(inputs, w) + b
        if activation == 'relu':
            out = tf.nn.relu(out, name=scope.name)
        elif activation == 'sigmoid':
            out = tf.nn.sigmoid(out, name=scope.name)

        return out
`
